File: frontend/lib/protocol.dart
enum TcpCommand {
  leftClick,
  rightClick,
  doubleClick,
  altTabStart,
  altTabEnd,
  volumeUp,
  volumeDown,
}

String formatUdpMove(double dx, double dy) => 'M,${dx.toStringAsFixed(1)},${dy.toStringAsFixed(1)}';
String formatUdpScroll(double dx, double dy) => 'S,${dx.toStringAsFixed(1)},${dy.toStringAsFixed(1)}';

String formatTcpCommand(TcpCommand command) {
  switch (command) {
    case TcpCommand.leftClick:
      return 'left_click';
    case TcpCommand.rightClick:
      return 'right_click';
    case TcpCommand.doubleClick:
      return 'double_click';
    case TcpCommand.altTabStart:
      return 'alt_tab_start';
    case TcpCommand.altTabEnd:
      return 'alt_tab_end';
    case TcpCommand.volumeUp:
      return 'volume_up';
    case TcpCommand.volumeDown:
      return 'volume_down';
  }
}

const udpPort = 8001;
const tcpPort = 8002;

File: backend/src/protocol.rs
use anyhow::{anyhow, Result};

#[derive(Debug, PartialEq)]
pub enum UdpPacket {
    Move { dx: f32, dy: f32 },
    Scroll { dx: f32, dy: f32 },
}

#[derive(Debug, PartialEq, Eq)]
pub enum TcpCommand {
    LeftClick,
    RightClick,
    DoubleClick,
    AltTabStart,
    AltTabEnd,
    VolumeUp,
    VolumeDown,
}

pub fn parse_udp_payload(payload: &[u8]) -> Result<UdpPacket> {
    let payload = std::str::from_utf8(payload)?.trim();
    let mut parts = payload.split(',');
    let kind = parts.next().unwrap_or_default();

    match kind {
        "M" => Ok(UdpPacket::Move {
            dx: parse_f32(parts.next())?,
            dy: parse_f32(parts.next())?,
        }),
        "S" => Ok(UdpPacket::Scroll {
            dx: parse_f32(parts.next())?,
            dy: parse_f32(parts.next())?,
        }),
        other => Err(anyhow!("unsupported UDP packet: {other}")),
    }
}

pub fn parse_tcp_payload(payload: &[u8]) -> Result<TcpCommand> {
    let payload = std::str::from_utf8(payload)?.trim();
    match payload {
        "left_click" => Ok(TcpCommand::LeftClick),
        "right_click" => Ok(TcpCommand::RightClick),
        "double_click" => Ok(TcpCommand::DoubleClick),
        "alt_tab_start" => Ok(TcpCommand::AltTabStart),
        "alt_tab_end" => Ok(TcpCommand::AltTabEnd),
        "volume_up" => Ok(TcpCommand::VolumeUp),
        "volume_down" => Ok(TcpCommand::VolumeDown),
        other => Err(anyhow!("unsupported TCP command: {other}")),
    }
}

fn parse_f32(value: Option<&str>) -> Result<f32> {
    let value = value.ok_or_else(|| anyhow!("missing delta value"))?;
    value.trim().parse::<f32>().map_err(Into::into)
}
